lib/csv/preprocessing.ts

```diff
@@ -0,0 +1,140 @@
+import DataFrame from 'dataframe-js'
+
+const uploadTypeProcessFunctionMap = {
+  [DataUploadEnum.APPLICANT]: processApplicantData,
+  [DataUploadEnum.TMUA_SCORES]: processTMUAData,
+  [DataUploadEnum.USER_ROLES]: processUserData
+}
+
+/**
+ * Parses a CSV and runs processing function according to the upload type
+ * @param file - the CSV file to process
+ * @param uploadType - the type of data being uploaded
+ * @returns flag to indicate success with data if true and errorMessage if false
+ */
+export async function preprocessCsvData(
+  file: File,
+  uploadType: DataUploadEnum
+): Promise<{ success: true; data: unknown[] } | { success: false; errorMessage: string }> {
+  if (file.name.split('.').pop() !== 'csv')
+    return { success: false, errorMessage: 'File must be a CSV' }
+  const lines = await file.text()
+
+  let objects: unknown[]
+  try {
+    objects = parse(lines, {
+      columns: true,
+      skip_empty_lines: true
+    })
+  } catch (e: any) {
+    if (e instanceof CsvError) {
+      return { success: false, errorMessage: `Error reading CSV: ${e.message}` }
+    }
+    return { success: false, errorMessage: 'Unexpected parsing error occurred.' }
+  }
+
+  try {
+    objects = uploadTypeProcessFunctionMap[uploadType](objects)
+  } catch (e: any) {
+    console.error(`error in CSV preprocessing: ${e.message}`)
+    return {
+      success: false,
+      errorMessage:
+        'Failure processing CSV: check the correct upload type has been selected and that all columns are named properly'
+    }
+  }
+
+  return { success: true, data: objects }
+}
+
+/**
+ * Process the data from the College-given applicant CSV file.
+ * Renames, transforms and select columns as appropriate.
+ * @param objects
+ * @returns an array of objects with nested applicant and application objects
+ */
+function processApplicantData(objects: unknown[]): unknown[] {
+  let df = new DataFrame(objects)
+  df = df.replace('', null)
+
+  const columnsToRename = [
+    ['College ID (Applicant) (Contact)', 'cid'],
+    ['UCAS ID (Applicant) (Contact)', 'ucasNumber'],
+    ['First name (Applicant) (Contact)', 'firstName'],
+    ['Last name (Applicant) (Contact)', 'surname'],
+    ['Nationality (Applicant) (Contact)', 'primaryNationality'],
+    ['Primary email address (Applicant) (Contact)', 'email'],
+    ['Gender (Applicant) (Contact)', 'gender'],
+    ['Date of birth (Applicant) (Contact)', 'dateOfBirth'],
+    ['Preferred first name', 'preferredName'],
+    ['Disability flag', 'hasDisability'], // TODO, we do not have a column in the CSV for this as of now
+    ['Entry term', 'admissionsCycle'],
+    ['Fee status', 'feeStatus'],
+    ['WP flag', 'wideningParticipation'],
+    ['Sent to department', 'applicationDate']
+  ]
+  columnsToRename.forEach(([oldName, newName]) => {
+    df = df.rename(oldName, newName)
+  })
+
+  // transform admissionsCycle column to a number
+  // ts-ignore is used because the types do not match the documentation or implementation
```

Comment:
What types are you referring to here? `string` and `DataFrame`?

Reply:
The `DataFrame.withColumn()` method has the wrong type of function in the documentation and types

Reply:
Clarified in e1b9ea86812472ac8254a641ee6496fc6e4cda70